src/providers/completer/environment.ts

```diff
@@ -83,14 +83,24 @@ export class Environment {
     }
 
 
-    provide(): vscode.CompletionItem[] {
+    provide(args: {document: vscode.TextDocument, position: vscode.Position}): vscode.CompletionItem[] {
         if (vscode.window.activeTextEditor === undefined) {
             return []
         }
         let snippetType: EnvSnippetType = EnvSnippetType.ForBegin
         if (vscode.window.activeTextEditor.selections.length > 1) {
             snippetType = EnvSnippetType.AsName
+        } else {
+            // If a closing '}' after '\begin{' has already been inserted, we need to remove it as it is already included in the snippets
+            const word = args.document.lineAt(args.position).text.slice(args.position.character - '\\begin{'.length, args.position.character + 2)
+            if (word === '\\begin{}') {
+                vscode.window.activeTextEditor.edit(e => {
+                    e.delete(new vscode.Range(args.position, args.position.translate(0, 1)))
+                })
+            }
+
```

Comment:
We should use `CompletionItem.range`. 

Calling `TextEditor.edit`, a function with side effects, leads to a bug like #1939.

Reply:
Good point! I had completely forgotten these side effect issues. I had already thought of `CompletionItem.range` but it forces us to update the every item in the suggestion list every time completion is triggered. I wanted to avoid that but I agree with you that this is probably the safest way to proceed.

Reply:
The new implementation uses `CompletionItem.range`.